tools/config.py

```diff
@@ -489,28 +502,90 @@ def regions(self):
         except KeyError:
             raise ConfigException("Not enough information in CMSIS packs to "
                                   "build a bootloader project")
+
+        rom_current_addr = 0
+        FLASH_ERASE_SIZE = self.target.flash_erase_size
+        VECTOR_TABLE_SIZE = self.target.vector_table_size
+
+        # Bootloader and Firmware Metadata Regions
         if 'target.bootloader_img' in target_overrides:
-            filename = target_overrides['target.bootloader_img']
-            part = intelhex_offset(filename, offset=rom_start)
-            if part.minaddr() != rom_start:
+            # find bootloader starting point
+            bootloader_start = rom_start
+
+            # find bootloader size
+            bootloader_fn = target_overrides['target.bootloader_img']
+            bootloader_part = intelhex_offset(bootloader_fn, offset=rom_start)
+            if bootloader_part.minaddr() != rom_start:
                 raise ConfigException("bootloader executable does not "
                                       "start at 0x%x" % rom_start)
-            part_size = (part.maxaddr() - part.minaddr()) + 1
-            yield Region("bootloader", rom_start + start, part_size, False,
-                         filename)
-            start += part_size
+            bootloader_size = (bootloader_part.maxaddr() - bootloader_part.minaddr()) + 1
+
+            # yeild bootloader region
+            yield Region("bootloader", bootloader_start, bootloader_size, False, bootloader_fn)
+
+            # update rom pointer
+            rom_current_addr = bootloader_start + bootloader_size
+
+            # find metadata header region start
+            metadata_start = rom_current_addr
+            if metadata_start%FLASH_ERASE_SIZE: # align to erase boundary
+                metadata_start = ((metadata_start // FLASH_ERASE_SIZE) + 1) * FLASH_ERASE_SIZE
+
+            # find metadata header size
+            metadata_size = FIRMWARE_HEADER_SIZE
+
+            # yeild image_metadata_header region
+            yield Region("firmware_metadata_header", metadata_start, metadata_size, False, None)
```

Comment:
We may not want to generate a firmware header for a particular pairing of bootloader and application. This does not support that use case. Let's make that possible.